### PR TITLE
Support Instantiating Choice Templates

### PR DESCRIPTION
**Pull request by Danaozhong**
Choices were so far greatly neglected, mainly because there are not frequently used.

This PR fixes the following:
- So far, template instantiation of compound types only  worked for structs. This PR adds support for instantiating templated choices.
- The default case of choices was completely ignored so far. This PR adds a) capitalization of the default case, b) the default case is added in the Marshal/Unmarshal/BitSize functions, c) the field member for the default choice now gets generated.
- The branching for choices was changed from a series of `if()` to a `switch:case`. The generated code not only looks cleaner, but should also run a tad faster. 
- A bug causing a segfault when passing parameters to optional fields is also fixed. This bug does not only affect choices, but all compound types (structs, enums, choices).

### CODE REVIEW

File: internal/generator/templates/choice.go.tmpl [hunk 1]
```diff
@@ -25,50 +30,76 @@ func (v *{{ $choice.Name}}) LoadDefaultValues() error {
...
-  if v.{{ $expr.ResultSymbol.Symbol.Name }} == {{ (index .Conditions 0).Condition.ResultIntValue }} { 
+  case {{ (index .Conditions 0).Condition.ResultIntValue }}: 
     {{- template "decode.go.tmpl" dict "pkg" $scope "field" .Field }}
-  }
   {{- end }}
 {{- end }}
-    _ = err // to avoid "declared but not used" warning
-    return nil
+{{- if $choice.DefaultCase }}
+  {{- if $choice.DefaultCase.Field }}
```
Comment (wichert) [inline internal/generator/templates/choice.go.tmpl:75]: Same as above

**Review by wichert — APPROVED** [review]
Looks good at first sight. It would be good to simplify the nested `if` statements in the templates, but not essential.

File: internal/model/transform.go [hunk 2]
```diff
@@ -210,6 +216,45 @@ func (m *Model) InstantiateTemplates() error {
...
+			}
+		}
+
+		tt, err := m.instantiateType(pkgScope, field.Type, newTemplateParameters, "")
+		if err != nil {
+			return nil, err
+		}
+		fieldCopy.Type = tt
+		return &fieldCopy, nil
+	} else if tt := templateTypes[field.Type.Name]; tt != nil && field.Type.Package == "" {
```
Comment (aignas) [inline internal/model/transform.go:245]: Consider following `if` statement guidelines from the effective go: https://go.dev/doc/effective_go#if as we can simplify things here.

What is more, if we have a map, consider using the `tt, ok := templateTypes[field.Type.Name]` pattern. This may still yield into a panic if there are `nil` items inside the map, but if that were the case, then we would know we have a much bigger problem.

File: internal/model/transform.go [hunk 3]
```diff
@@ -326,6 +339,113 @@ func (m *Model) instantiateStruct(
...
+		}
+		templateTypes[templChoice.TemplateParameters[ix]] = ta
+	}
+
+	newType := &ast.TypeReference{
+		IsBuiltin: false,
+		Package:   pkgScope.Name,
+		Name:      instantiatedName,
+	}
+	if _, exists := pkgScope.Choices[newType.Name]; !exists {
```
Comment (aignas) [inline internal/model/transform.go:370]: It seems that we can invert an if here. The code inside the statement is a bit long and it becomes hard to read this way.

File: internal/model/transform.go [hunk 4]
```diff
@@ -326,6 +339,113 @@ func (m *Model) instantiateStruct(
...
+		choice := &ast.Choice{
+			Name:           newType.Name,
+			Comment:        templChoice.Comment,
+			Cases:          []*ast.ChoiceCase{},
```
Comment (aignas) [inline internal/model/transform.go:374]: nit: We could in theory leave this out as append will work with a `nil` value in the same way. The same for the `TypeParameters` field.

File: internal/model/transform.go [hunk 5]
```diff
@@ -326,6 +339,113 @@ func (m *Model) instantiateStruct(
...
+				if err != nil {
+					return nil, err
+				}
+				choice.DefaultCase.Field = newField
+			}
+
+		}
+		// resolve the templates for the types
+		for _, typeParam := range templChoice.TypeParameters {
+			if tt := templateTypes[typeParam.Type.Name]; tt != nil && typeParam.Type.Package == "" {
```
Comment (aignas) [inline internal/model/transform.go:413]: similar to my comment above about handling of maps.

**Review by aignas — COMMENTED** [review]
Whenever we are returning errors, it would be great to use `fmt.Errorf("some context: %w", err)` instead of just `return err`. That way any error is much easier to debug and error messages out of the CLI can become easier to read.

And I know that it is difficult to test the `model` package, but it would be great to start adding unit tests in here.

File: internal/model/transform.go (hunk 2 above)
Comment (Danaozhong) [reply internal/model/transform.go]: done

File: internal/model/transform.go (hunk 5 above)
Comment (Danaozhong) [reply internal/model/transform.go]: done

File: internal/model/transform.go (hunk 4 above)
Comment (Danaozhong) [reply internal/model/transform.go]: done

File: internal/generator/templates/choice.go.tmpl [hunk 6]
```diff
@@ -25,50 +30,76 @@ func (v *{{ $choice.Name}}) LoadDefaultValues() error {
...
-  if v.{{ $expr.ResultSymbol.Symbol.Name }} == {{ (index .Conditions 0).Condition.ResultIntValue }} { 
+  case {{ (index .Conditions 0).Condition.ResultIntValue }}:
     {{- template "encode.go.tmpl" dict "pkg" $scope "field" .Field }}
-  }
   {{- end }}
 {{- end }}
-    _ = err // to avoid "declared but not used" warning
-    return nil
+{{- if $choice.DefaultCase }}
+  {{- if $choice.DefaultCase.Field }}
```
[the remark this answers is not in the document]
Comment (Danaozhong) [reply internal/generator/templates/choice.go.tmpl:53]: done

File: internal/model/transform_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+					test.testPackage.Name: test.testPackage,
+				},
+			}
+			if err := test.testPackage.CollectSymbols(); err != nil {
+				t.Fatalf("error during symbol collection: %v", err)
+			}
+
+			if err := m.InstantiateTemplates(); err != nil {
+				t.Fatalf("error during template instantiation: %v", err)
+			}
```
Comment (aignas) [inline internal/model/transform_test.go:96]: nit: one could use the `testify/require` package to shorten this to:
```golang
require.NoError(t, m.InstantiateTemplates(), "failed to instantiate templates")
```

File: internal/model/transform_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+			// search for the instantiated struct
+			if s, ok := test.testPackage.Structs[test.resultingType.Name]; !ok {
+				t.Fatalf("instantiated struct not found : %s", test.resultingType.Name)
+			} else {
+				if diff := cmp.Diff(test.resultingType, s); diff != "" {
+					t.Errorf("template instantiation does not match: %s", diff)
+				}
+			}
```
Comment (aignas) [inline internal/model/transform_test.go:105]: And this one to:
```golang
assert.Contains(t, test.resultingType, test.testPackage.Structs)
```

**Review by aignas — COMMENTED** [review]
Thank you for the cleanup!
